code/controllers/subsystems/ai.dm

```diff
@@ -23,14 +24,28 @@ SUBSYSTEM_DEF(ai)
 		if (!length(queue))
 			return
 	var/cut_until = 1
+	#ifdef INCLUDE_URIST_CODE
+	var/throttle_on_empty = prob(config.run_empty_levels_throttled_perc) // Urist edit!
+	#endif
 	for (var/datum/ai_holder/ai as anything in queue)
 		++cut_until
 		if (QDELETED(ai) || ai.busy)
 			continue
 		if (!ai.holder)
 			continue
+		#ifndef INCLUDE_URIST_CODE
 		if (!config.run_empty_levels && !SSpresence.population(get_z(ai.holder)))
 			continue
+		#else
+		// Allow 'leaky' processing of empty, activated z-levels
+		if (!config.run_empty_levels)
+			var/holder_z = get_z(ai.holder)
+			var/zlevel_has_pop = SSpresence.population(holder_z)
+			if(!zlevel_has_pop)
+				var/zlevel_had_pop = SSpresence.population_from_cache(holder_z)
+				if(!zlevel_had_pop || (zlevel_had_pop && throttle_on_empty))
```

Comment:
Extra `zlevel_has_pop` is redundant as we've already checked the state in the previous OR branch which is lazily evaluated.

Would it also be possible to factor in populated connected z-levels in here too? Something kinda like...

```suggestion
				var/zlevel_had_pop = SSpresence.population_from_cache(holder_z)
				var/connected_has_pop = FALSE
				for(var/connected_z in GetConnectedZlevels(holder_z))
					if(SSpresence.population(connected_z))
						connected_has_pop = TRUE
						break
				if((!zlevel_had_pop && !connected_has_pop) || throttle_on_empty)
```

Reply:
Best I can recall, I've done it this way just to be explicit about the logic. 

For connected Zs, I'd like to test it out with the leaky processing on one Z first. My concern is that this might heavily de-optimize the current LODs in the worst case where players are heavily scattered across levels.

Reply:
Could we not edit `SSpresence::population_from_cache()` to also cache over connected Z levels then? Ie for all connected z-levels, store in the cache that they have previously been populated?

Not a big deal, but it'd be nice for instances like the Nerva if no-one has been on deck 1, but to still allow AI on that level to process at a reduced rate